_published/0016-wharf-api-endpoints-cleanup.md

```diff
@@ -0,0 +1,888 @@
+---
+layout: default
+# This is just to fool remark-stringify not to escape & symbols
+# See https://github.com/syntax-tree/mdast-util-to-markdown/issues/8
+shields_io_query_params: label=issue%20state&logo=github&style=flat-square
+
+# Update the following (it's YAML syntax)
+pr_id: 16 # Update this with PR number/ID. No leading zeros
+rfc_feature_name: wharf-api-endpoints-cleanup # Use kebab-case
+title: "RFC-0016: wharf-api-endpoints-cleanup" # Update this with PR number/ID and feature name. Use leading zeros
+rfc_author_username: jilleJr
+rfc_author_name: Kalle Jillheden # Or same as username, if you wish
+
+# Leave these. Collaborator changes this before merging
+impl_issue_id: 0
+impl_issue_repo: iver-wharf/wharf-api
+last_modified_date: YYYY-MM-DD
+---
+
+# {{page.title}}
+
+- RFC PR: [iver-wharf/rfcs#{{page.pr_id}}](https://github.com/iver-wharf/rfcs/pulls/{{page.pr_id}})
+- Feature name: `{{page.rfc_feature_name}}`
+- Author: {{page.rfc_author_name}} ([@{{page.rfc_author_username}}](https://github.com/{{page.rfc_author_username}}))
+- Implementation issue: [{{page.impl_issue_repo}}#{{page.impl_issue_id}}](https://github.com/{{page.impl_issue_repo}}/issues/{{page.impl_issue_id}})
+- Implementation status: ![GitHub issue state](https://img.shields.io/github/issues/detail/state/{{page.impl_issue_repo}}/{{page.impl_issue_id}}?{{page.shields_io_query_params}})
+
+## Summary
+
+Restructuring the wharf-api by changing the paths of the endpoints, adding and
+renaming path parameters, and changing the request and response models around.
+
+While we will keep backward compatibility for at least one major version, we
+will be changing so much that this needs a full major version bump.
+
+This is not the full list of changes to be made for v5. Instead, it's
+a list of changes that will require us moving to it. Changes not listed in
+this RFC may be included in the version bump as well.
+
+## Motivation
+
+We are lacking consistency in our API, and there are some questionable
+endpoints in there. Most annoying issues are:
+
+- Some endpoints are plural, some are singular.
+  Ex: `POST /project` vs `GET /projects`.
+
+- Some are misplaced and rely on request body instead of path/query
+  parameters. Ex: `PUT /branches` instead of `PUT /project/{projectId}/branch`.
+
+- Path parameters are all lowercased, while all other query parameters are
+  camelCased.
+
+- We use the database models in our request and response specs. This is a
+  security weakness as it induces a higher risk of exposing too much data to
+  the user. With separate models we would also make the Swagger documentation
+  easier to understand, where we could elide irrelevant specifications such as
+  allowing to specify a project ID when creating a new project.
```

Comment:
I think the strongest argument that you've missed is that this will hugely improve decoupling (i.e. easier to change the database model without changing the API model, making changes less viral) :)

Reply:
Gosh! Probably in the back of my mind as I was writing it, but thanks for reminding me!